erizo/src/erizo/rtp/QualityManager.cpp

```diff
@@ -4,7 +4,112 @@ namespace erizo {
 
 DEFINE_LOGGER(QualityManager, "rtp.QualityManager");
 
-QualityManager::QualityManager()
-  : spatial_layer_{0}, temporal_layer_{0}, padding_enabled_{false} {}
+constexpr duration QualityManager::kMinLayerSwitchInterval;
+constexpr duration QualityManager::kActiveLayerInterval;
+
+QualityManager::QualityManager(std::shared_ptr<Clock> the_clock)
+  : initialized_{false}, padding_enabled_{false}, forced_layers_{false},
+  spatial_layer_{0}, temporal_layer_{0},
+  current_estimated_bitrate_{0}, last_quality_check_{the_clock->now()},
+  clock_{the_clock} {}
+
+
+void QualityManager::notifyQualityUpdate() {
+  if (!initialized_) {
+    auto pipeline = getContext()->getPipelineShared();
+    if (!pipeline) {
+      return;
+    }
+    stats_ = pipeline->getService<Stats>();
+    if (!stats_) {
+      return;
+    }
+    if (!stats_->getNode()["total"].hasChild("senderBitrateEstimation")) {
+      return;
+    }
+    initialized_ = true;
+  }
+  if (forced_layers_) {
+    return;
+  }
+  time_point now = clock_->now();
+  uint64_t estimated_bitrate = stats_->getNode()["total"]["senderBitrateEstimation"].value();
+  bool available_bitrate_is_descending = estimated_bitrate < current_estimated_bitrate_;
+  current_estimated_bitrate_ = estimated_bitrate;
+  if (!isInBaseLayer() &&  (available_bitrate_is_descending || !isCurrentLayerPresent())) {
```

Comment:
Instead of available_bitrate_is_descending, I'd call selectLayers when `current_estimated_bitrate_ < current_layer_bitrate_` because layer bitrate may ascend too.